clippy_lints/src/cognitive_complexity.rs

```diff
@@ -48,190 +52,329 @@ impl LintPass for CognitiveComplexity {
+                // in the current chain of logical operators.
+                //
+                // Say, for example, that we're scoring `a && b && c || d && e`.
+                // There are 2 changes in the operator chain in this expression,
+                // once at `c || d` (it changes from `&&` to `||`) and once at
+                // `d && e` (it changes from `||` to `&&`).
+                //
+                // In order for us to score this change, regardless of the shape of
+                // the AST, we need to be able to know which operator sits right
+                // next to the current one. If it's then a different operator,
+                // we know there is a change in the chain, and we can score it.
+                //
+                // But what about scoring it twice? Will we see the same change
+                // more than once?
+                // The answer is no: since ASTs are recursive, child operators
+                // can't see their parent operators. Given we're only scoring
+                // a change whenever the operators right next to the current one
+                // are different to it, AND in our subsequent calls the current
+                // operator will not be visible, it's effectively impossible
+                // to score this change in the chain more than once.
+                enum At {
+                    LeftOfRightSide,
+                    RightOfLeftSide,
+                }
+
+                // A logical operator
+                #[derive(PartialEq)]
+                enum LogOp {
+                    LogAnd, // &&
+                    LogOr,  // ||
+                    BitAnd, // &
+                    BitOr,  // |
+                    BitXor, // ^
+                    None,   // Other
+                }
+
+                // Translate from a binary operator to a logical operator
+                fn log_op_from_bin_op(bop_kind: BinOpKind) -> LogOp {
+                    match bop_kind {
+                        BinOpKind::And => LogOp::LogAnd,
+                        BinOpKind::Or => LogOp::LogOr,
+                        BinOpKind::BitAnd => LogOp::BitAnd,
+                        BinOpKind::BitOr => LogOp::BitOr,
+                        BinOpKind::BitXor => LogOp::BitXor,
+                        _ => LogOp::None,
+                    }
+                }
+
+                /// Find the rightmost or leftmost logical operator inside of
+                /// the given AST's left side or right side, respectively.
+                fn log_op_at(expr: &Expr, at: At) -> LogOp {
+                    match &expr.node {
+                        ExprKind::Binary(binop, ref left_side, ref right_side) => {
+                            let current_operator = log_op_from_bin_op(binop.node);
+
+                            let next_operator = match at {
+                                At::LeftOfRightSide => log_op_at(right_side, at),
```

Comment:
I think you swapped these :D

Reply:
Whoops, you're right. Or at least, there's an error in the function. In its current state, it doesn't recurse correctly. I think the previous version (the non-general one) did the same.

Reply:
It's fixed now :D

Reply:
There's also a test for it now